### DISCUSSION TITLE
Lens and Mirror together in one optical system

### DISCUSSION THREAD

**Discussion by PoulomiPradhan** [mjhoptics/ray-optics#130] [Q&A] [answered]
I would like to simulate an optical system which consists of a spherical lens whose one side is a mirror. It would be great if I get any help on how I can simulate the same.

**Comment by mjhoptics** [comment, marked as answer]
## How to simulate an optical system which consists of a spherical lens whose one side is a mirror.


```python
%matplotlib inline
isdark = False
from rayoptics.environment import *
```


```python
opm = OpticalModel()
sm  = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']
```


```python
osp['pupil'] = PupilSpec(osp, key=['object', 'pupil'], value=100.)
osp['fov'] = FieldSpec(osp, key=['object', 'angle'], value=8.0, flds=[0., 1.], is_relative=True)
osp['wvls'] = WvlSpec([('F', 0.5), ('d', 1.0), ('C', 0.5)], ref_wl=1)
```

```python
opm.radius_mode = True

sm.gaps[0].thi=1e10
```

The front, transmitting surface of the lens is entered twice: surfaces 1 and 3
The second, mirrored lens surface is entered once, as a `refl` surface.


```python
sm.add_surface([-1200., 10., 'N-BK7', 'Schott'])
sm.add_surface([-1000., -10., 'refl'])
sm.add_surface([-1200., -450.7])
```


```python
opm.update_model()
```

## Draw a lens picture


```python
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, 
                        do_draw_beams=False, do_draw_edge_rays=True,
                        do_draw_ray_fans=False, num_rays_in_fan=7, is_dark=isdark).plot()
```


    
![output_9_0](https://github.com/mjhoptics/ray-optics/assets/33294031/ba9464da-f29d-4b45-9c09-ff6a65bf13e5)
    



```python
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1  1.4054e+09
        1: -1200.000000      10.0000     N-BK7             1      50.147
        2: -1000.000000     -10.0000     N-BK7    reflect -1      50.892
        3: -1200.000000     -450.700       air            -1      50.935
      Img:    -0.000000      0.00000                      -1      66.067



```python
listobj(osp)
```

    aperture: object epd; value=100.0
    field: object angle; value=8.0
    x,y=0.0 vlx= 0.000 vux= 0.000 vly= 0.000 vuy= 0.000
    y=1.0 vly= 0.000 vuy= 0.000 vlx= 0.000 vux= 0.000
    central wavelength=  587.5618 nm
    wavelength (weight) =  486.1327 (0.500),  587.5618 (1.000)*,  656.2725 (0.500)
    focus shift=0.0
    



```python
pm.first_order_data()
```

    efl              -457.2
    f                 457.2
    f'               -457.2
    ffl              -450.7
    pp1               6.574
    bfl              -450.7
    ppk               6.574
    pp sep        5.063e-14
    f/#               4.572
    m            -4.572e-08
    red          -2.187e+07
    obj_dist          1e+10
    obj_ang               8
    enp_dist             -0
    enp_radius           50
    na obj            5e-09
    n obj                 1
    img_dist         -450.7
    img_ht            64.26
    exp_dist          13.29
    exp_radius        50.73
    na img           0.1087
    n img                -1
    optical invariant        7.027



```python
abr_plt = plt.figure(FigureClass=RayFanFigure, opt_model=opm, data_type='Ray', 
                     scale_type=Fit.All_Same, is_dark=isdark).plot()
```


    
![output_13_0](https://github.com/mjhoptics/ray-optics/assets/33294031/74490ae2-1af3-4c50-b72b-e3e18808f5c6)

**Comment by PoulomiPradhan** [comment]
Thank you very much!
